README.md
## freezeframe.js 3.0.10

freezeframe.js is a script that pauses animated .gifs and enables them to 
animate on mouse hover / mouse click / touch event, or with trigger / release 
functions. It supports responsive images and works as a jQuery / Zepto plugin.

1. [Examples](#examples)
2. [Files & Dependencies](#files_dependencies)
3. [HTML Setup](#html_setup)
4. [Basic Usage: jQuery / Zepto Plugin](#basic_usage_jquery_zepto_plugin)
5. [Basic Usage: Vanilla JS](#basic_usage_vanilla_js)
6. [Advanced Usage](#advanced_usage)
7. [Options Reference](#options_reference)
8. [Function Reference](#function_reference)
9. [License](#license)
10. [Thanks](#thanks)

<a name="examples"></a>
## Examples
[http://ctrl-freaks.github.io/freezeframe.js/](http://ctrl-freaks.github.io/freezeframe.js/)

<a name="files_dependencies"></a>
## Files & Dependencies

1. Include the js and css from **/build/**  

```
  <link rel="stylesheet" href="freezeframe_styles.min.css">
  <script src="freezeframe.min.js"></script>
    ```
2. If you do not use the **packaged version** of freezeframe, the following dependencies are required:
  * imagesLoaded Packaged v4.1.4 ( [https://github.com/desandro/imagesloaded](https://github.com/desandro/imagesloaded) )
  * jQuery v2.1.4 *or* Zepto v1.1.6  
    (**Note**: only the core and event modules are required if using Zepto, 20kb minified)


<a name="html_setup"></a>
## HTML Setup

Add **freezeframe** as a class name on the .gifs you want processed.  
( You can optionally specify a custom selector as shown in [Advanced Usage](#advanced_usage). )

```HTML5
<img class="freezeframe" src="image.gif" /> 
```
Add **freezeframe-responsive** as an additional class name to make the .gif responsive.

```HTML5
<img class="freezeframe freezeframe-responsive" src="image.gif" /> 
```

<a name="basic_usage_vanilla_js"></a>
## Basic Usage: Vanilla JS

✨Freeze those frames ✨

```javascript
$(function() {
  ff = new freezeframe().freeze();
})
```

<a name="basic_usage_jquery_zepto_plugin"></a>
## Basic Usage: jQuery / Zepto Plugin

Trigger the image you want by selector and run the **freezeframe()** function. 

```javascript
$('.my_class').freezeframe();
```

<a name="advanced_usage"></a>
## Advanced Usage

freezeframe.js exposes public methods to allow for more custom integration. You 
have the option of manually controlling when freezeframe triggers images, adds 
support elements, and attaches event handlers. You can also manually trigger 
and release animation on one image or a group of images. These methods are 
described in detail in the [Function Reference](#function_reference).

*Example: trigger logo .gif and manually trigger / release animation:*
```javascript
logo = new freezeframe('#logo'); // setup freezeframe instance with custom selector

logo.trigger(); // trigger images by selector
logo.setup();   // setup support elements
logo.trigger(); // trigger animation
logo.release(); // release animation
```

<a name="options_reference"></a>
## Options Reference

Options can be passed to a freezeframe instance on creation in the form of an 
**object** or a **string**. Strings will be interpreted as the **selector** option.  

```javascript
// Options passed to jQuery plugin
$('.my_class').freezeframe({'animation_play_duration': 2500})

// String as selector option
var ff = new freezeframe('.my_class');

// Object as options
var ff = new freezeframe({
  'selector': '.my_class',
  'animation_play_duration': 3000,
  'non_touch_device_trigger_event': 'hover'
})
```

* ### ```selector```   

    <code><b>type:</b> string</code>  
    <code><b>default</b>: ".freezeframe"</code>  

    The selector used to search for .gifs when the ```trigger()``` function is 
    run.

* ### ```animation_play_duration```  

    <code><b>type:</b> integer</code>  
    <code><b>default</b>: 5000</code>  
    <code><b>options</b>: Infinity</code>  

    The number of milliseconds a .gif will animate for when triggered by click / 
    touch event.  
    Use ```Infinity``` to make the .gif play *forever.*

* ### ```non_touch_device_trigger_event```  

    <code><b>type:</b> string</code>  
    <code><b>default</b>: "hover"</code>  
    <code><b>options</b>: "hover", "click"</code>  

    The trigger event to start animation for non-touch devices.

<a name="function_reference"></a>
## Function Reference

* ### ```freezeframe( options )```  

    Create a new freezeframe object instance.  
    Can be passed options. Strings will be interpreted as the **selector** option. 
     ```javascript
  // Default options
  var ff = new freezeframe();

  // String as selector option
  var ff = new freezeframe('.my_class');

  // Object as options
  var ff = new freezeframe({
    'selector': '.my_class',
    'animation_play_duration': 3000,
    'non_touch_device_trigger_event': 'hover'
  })
    ```

* ### ```trigger( selector )```  
    trigger images to be paused by freezeframe. If run without selector 
    argument, selector in freezeframe options will be used. Can be run multiple 
    times with different selector to group many images, unrelated by selector, 
    in one freezeframe instance.
    ```javascript
  ff.trigger();               // use selector in freezeframe options
  ff.trigger('.my_class');    // use custom selector
    ```

* ### ```setup()```  
    Creates and inserts support elements. Can be filtered by selector.  
    ```javascript
  ff.setup();                 // all images in freezeframe instance
  ff.setup('.my_class');      // filter images by selector
    ```
    HTML image before:

     ```
  <img class="freezeframe" src="my_image.gif" />
     ```  

     ...and after:
     ```
  <div class="ff-container">
    <canvas class="ff-canvas ff-canvas-ready" width="400" height="250"></canvas>
    <img class="freezeframe ff-setup ff-image ff-image-ready" src="my_image.gif">
  </div>
     ```

* ### ```attach( selector )```  
    Attaches hover / click / touch events based on freezeframe options. Can be filtered by selector.
    ```javascript
  ff.attach();                // all images in freezeframe instance
  ff.attach('.my_class');     // filter images by selector
    ```

* ### ```trigger( selector )```  
    Triggers (starts) animation, or restarts animation from the first frame if 
    the .gif is already animating. Can be filtered by selector.
    ```javascript
  ff.trigger();               // all images in freezeframe instance
  ff.trigger('.my_class');    // filter images by selector
    ```

* ### ```release( selector )```  
    Releases (stops) animation. Can be filtered by selector.
    ```javascript
  ff.release()                // all images in freezeframe instance
  ff.release('.my_class');    // filter images by selector
    ```

<a name="license"></a>
## License
freezeframe.js is released under the terms of the MIT License.

<a name="thanks"></a>
## Thanks
* [Browserstack - Live, Web-Based Browser Testing](https://www.browserstack.com/)


CHANGELOG.md

# Change Log
All notable changes to this project will be documented in this file.  
This project adheres to [Semantic Versioning](http://semver.org/).

## [3.0.10] - 2018-05-25
### Fied
- Fixed mobile issue

## [3.0.9] - 2018-05-21
### Added
- Integrated overlay icon feature

## [Unreleased]
### Added
- Zepto file to src/js/vendor
### Removed
- jQuery dependency
- jQuery file from src/js/vendor
## Changed
- jQuery hard dependency to Zepto or jQuery

## [3.0.8] - 2016-05-03
### Added
- Semver versioning re-introduced
- Changelog updated to match format propsed by [http://keepachangelog.com/](http://keepachangelog.com/)

## [3.0.7] - 2016-04-27
### Fixed
- Core methods are now exposed to the jQuery object

## [3.0.6] - 2016-04-11
### Changed
- Attribution-ShareAlike 3.0 to MIT License

## [3.0.5] - 2016-03-17
### Added
- Thanks to [Browserstack](http://www.browserstack.com/) to the readme for their sponshorship!

## [3.0.4] - 2016-02-11
### Added
- Infinity as a value for animation duration

## [3.0.3] - 2016-02-07
### Added
- Private methods to protect some parts the user shouldnt be touching  

### Fixed
- Trigger / release functions so they don't fire if the selector doesn't match 
any captured images

## [3.0.2] - 2016-02-01
### Added
- Responsive image support

## [3.0.1] - 2016-01-23
### Fixed
- Documentation updated and general code cleanup

## [3.0.0] - 2016-01-09
### Added
- Totally new code! Complete rewrite  

### Removed
- All the old code!  

### Fixed
- Broke the script down into smaller functions
- Simplified input options

## [2.0.5] - 2015-01-15
### Fixed
- Added '?' to extension when filtering out non .gif files to allow for images 
with file extensions that prevent caching such as my_image.gif?2384792384723

## [2.0.4] - 2014-09-10
### Fixed
- Typo in support_touch_devices variable

## [2.0.3] - 2014-08-29
### Added
- Bower.json file

## [2.0.2] - 2014-02-09
### Fixed
- More jQuery no conflict issues

## [2.0.1] - 2014-01-07
### Fixed
- Bug with selecting images, number of classnames is no longer an issue

## [2.0.0] - 2014-30-1
### Added
- Attribution-ShareAlike 3.0 License  

### Changed
- Boolean ```freezeframe``` attribute to classname  

### Removed
- Broken debug functions and variables
- Ability to create freezeframe instance yourself, instead freezeframe just 
runs on page load

## [1.2.1] - 2013-08-17
### Added
- Cross origin image support

## [1.2.0] - 2013-03-27
### Fixed
- Bug with touch device detection
- Bug with jQuery .load not firing correctly for cached images
- Bug with jQuery conflicts
- Bug with dynamically resized images

### Added
- Using [isMouseEventSupported by kangax](http://perfectionkills.com/detecting-event-support-without-browser-sniffing/) 
for feature testing
- Using [imagesLoaded by desandro](https://github.com/desandro/imagesloaded)

## [1.1.0] - 2012-12-07
### Added
- Icon to represent animation and supporting variables
- Support for touch devices and supporting variables
- Ability to disable loading and animation icons

### Changed
- Variables for loading background style

## [1.0.2] - 2012-11-30
### Fixed
- Filename and version syntax

## [1.0.1] - 2012-11-28
### Fixed
- Image positioning bug

## [1.0.0] - 2012-11-25
### Added
- Initial release


freezeframe.js
/* -----------------------------------------------------------------------------
 * Freezeframe
 * freezeframe.js v2.0.2
 * 2014 Chris Antonellis
 *
 * Freezeframe.js is a script that automatically pauses animated GIFs and 
 * enables them to start animating on mouse hover.
 *
 * Website: http://freezeframe.chrisantonellis.com/
 * Documentation: http://freezeframe.chrisantonellis.com/documentation/
 * 
 * Licensed under a Creative Commons Attribution-ShareAlike 3.0 Unported License
 * http://creativecommons.org/licenses/by-sa/3.0/deed.en_US
 * -------------------------------------------------------------------------- */

FreezeFrame = (function($) {

	var _ff;
	var canvas;
	var context;

	var class_name;
	var trigger_event;
	var support_touch_devices;
	var animation_play_duration;

	var loading_background_color;
	var loading_background_image;
	var loading_background_position;
	var loading_fade_in_speed;

	var animation_icon_image;
	var animation_icon_position;
	var animation_fade_out_speed;

	var is_touch_device;
	var freezeframe_count;

	function FreezeFrame( _options ) {

		_ff = this;

		_options.class_name == null ?
			this.class_name = "freezeframe" : 
			this.class_name = _options.class_name;

		_options.trigger_event == null ? 
			this.trigger_event = "hover" : 
			this.trigger_event = _options.trigger_event.toLowerCase();

		_options.support_touch_devices == null ? 
			this.support_touch_devices = true : 
			this.support_touch_devices = _options.support_touch_devices.toLowerCase();

		_options.animation_play_duration == null ? 
			this.animation_play_duration = 10000 : 
			this.animation_play_duration = parseInt(_options.animation_play_duration);

		_options.loading_background_color == null ? 
			this.loading_background_color = "#666" : 
			this.loading_background_color = _options.loading_background_color.toLowerCase();

		_options.loading_background_image == null ? 
			this.loading_background_image = "data:image/gif;base64,R0lGODlhEAAQAPIAAGZmZv///4mJidbW1v///8PDw7CwsKampiH+GkNyZWF0ZWQgd2l0aCBhamF4bG9hZC5pbmZvACH5BAAKAAAAIf8LTkVUU0NBUEUyLjADAQAAACwAAAAAEAAQAAADMwi63P4wyklrE2MIOggZnAdOmGYJRbExwroUmcG2LmDEwnHQLVsYOd2mBzkYDAdKa+dIAAAh+QQACgABACwAAAAAEAAQAAADNAi63P5OjCEgG4QMu7DmikRxQlFUYDEZIGBMRVsaqHwctXXf7WEYB4Ag1xjihkMZsiUkKhIAIfkEAAoAAgAsAAAAABAAEAAAAzYIujIjK8pByJDMlFYvBoVjHA70GU7xSUJhmKtwHPAKzLO9HMaoKwJZ7Rf8AYPDDzKpZBqfvwQAIfkEAAoAAwAsAAAAABAAEAAAAzMIumIlK8oyhpHsnFZfhYumCYUhDAQxRIdhHBGqRoKw0R8DYlJd8z0fMDgsGo/IpHI5TAAAIfkEAAoABAAsAAAAABAAEAAAAzIIunInK0rnZBTwGPNMgQwmdsNgXGJUlIWEuR5oWUIpz8pAEAMe6TwfwyYsGo/IpFKSAAAh+QQACgAFACwAAAAAEAAQAAADMwi6IMKQORfjdOe82p4wGccc4CEuQradylesojEMBgsUc2G7sDX3lQGBMLAJibufbSlKAAAh+QQACgAGACwAAAAAEAAQAAADMgi63P7wCRHZnFVdmgHu2nFwlWCI3WGc3TSWhUFGxTAUkGCbtgENBMJAEJsxgMLWzpEAACH5BAAKAAcALAAAAAAQABAAAAMyCLrc/jDKSatlQtScKdceCAjDII7HcQ4EMTCpyrCuUBjCYRgHVtqlAiB1YhiCnlsRkAAAOwAAAAAAAAAAAA==" : 
			this.loading_background_image = _options.loading_background_image;

		_options.loading_background_position == null ? 
			this.loading_background_position = "center center" : 
			this.loading_background_position = _options.loading_background_position.toLowerCase();

		_options.animation_icon_image == null ? 
			this.animation_icon_image = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAl9JREFUeNrs2s2K2lAUAGATb3TE2Cr+oaJQNFOh1MVsxK0OuJbSdX2V9gG67GqW3RbaB7B9gy7qgKJF8G/EarHVWGMyJj230EXLLITm6q05F44/IVw5X3JPjiGCZVkuJw/R5fCBAAiAAAiAAAiAAAiAAAiAAAiAAAiAAAiAAAjwjyMej79MpVKP/hcAwe6boul0+p3P5/vmdruv4evbVqvVdRRALpd7HQqFpn6/XxUEwdI07UZV1ffNZvMzjwDE7gk9Hs8mGAwuYCl8hTNB32630nw+f1IqlRY8QtgOQAjRZVleJxKJ77AcVLptOBwup9PpPR4hbAcQRdGUJOk2HA5riqKokUjEgKK4GY/HKo8QhMGcv4oKIJg0+Xw+r8H7bSaT0XiEICwnhyLoCgQCptfrNSgCjxDkED8ChdGiwSMEOaQ2jxDkGOtuH4jJZHJ/Nps9LRaLi+Vy2WDVUB0FYB+IXq+36ff7P0ajURj2eQbF9EO73W6cFMBdENA8mXQbNFDier0mhULhQaVSeZzNZj+d3Bnwe+i6LkDCAqx9QmMwGJwlk8nzarV6Cf3EDHZ5DvHx5ADuSjwWiz0sl8uX8H9izjLxowLskfgL1okfBYCnxA8KwGPiBwGg9xpWq5XIY+IsAQT6stvt6BGX4Nrt4jFxZgCmaYqGYRDo4nzdbncny3KGx8SZAcCRl6CH9yuKcl6r1S6i0egXHhP/Y53aGfV6/U2n02nA5yuIC7vntzsEBo/KXkG84vaI/12w8Flhhw8EQAAEQAAEQAAEQAAEQAAEQAAEQAAEQAAEcNr4KcAABsB8S2vxAV8AAAAASUVORK5CYII=" :
			this.animation_icon_image = _options.animation_icon_image;

		_options.animation_icon_position == null ? 
			this.animation_icon_position = "top left" : 
			this.animation_icon_position = _options.animation_icon_position.toLowerCase();

		_options.loading_fade_in_speed == null ? 
			this.loading_fade_in_speed = 500 :
			this.loading_fade_in_speed = parseInt(_options.loading_fade_in_speed);

		_options.animation_fade_out_speed == null ? 
			this.animation_fade_out_speed = 250 : 
			this.animation_fade_out_speed = parseInt(_options.animation_fade_out_speed);

		this.is_touch_device = isMouseEventSupported("ontouchstart");
		this.freezeframe_count = 0;
	};

	/** --------------------------------------------------------------------------
	 * @function .setup()
	 * Sets up the freezeframe instance by adding a canvas element to the 
	 * document and capturing references to the canvas and its 2D context
	 * ------------------------------------------------------------------------ */

	FreezeFrame.prototype.setup = function() {
		$("<canvas>",{id:"freezeframe-canvas"})
			.css("display", "none")
			.prependTo($("body"));
		this.canvas = $("canvas#freezeframe-canvas");
		this.context = this.canvas[0].getContext('2d');
	};

	/** --------------------------------------------------------------------------
	 * .run()
	 * Starts freezeframe processing for each image. Copies the image to 
	 * the freezeframe canvas and converts it to a data url
	 * ------------------------------------------------------------------------ */

	FreezeFrame.prototype.run = function() {
		var images = [];
		var ext;
		var figure_background = _ff.loading_background_color;

		if( _ff.loading_background_image !== false ) {
			figure_background += " url('" + _ff.loading_background_image + "') " +
				_ff.loading_background_position + " no-repeat";
		}

		// Select all images with a class matching the option class_name
		images = $('img[class*="' + _ff.class_name + '"]')
			.not('[class*="' + _ff.class_name + '_done"]');

		// Process each image by resetting the animation sequence, copying to the 
		// canvas, converting to a data url, and attaching that data url to the 
		// image itself as an attribute
		$(images).each(function(index) {
			// Change image class so it won't be reprocessed if .run() is run again
			$(this).removeClass(_ff.class_name).addClass(_ff.class_name + "_done");
			// Set cross-origin to anon to load images from remote services that send 
			// the correct header. Not working correctly, needs more testing
			$(this).crossOrigin = "anonymous";
			// Determine file extension
			ext = $(this)[0].src.split(".");
			ext = ext[ext.length - 1].toLowerCase();
			ext = ext.split("?")[0];
			// Remove non GIF files
			if(ext !== "gif") {
				images.splice(index, 1);
			} else {

				var freezeframe_figure = $("<figure />")
					.attr("class", "freezeframe-container " + _ff.freezeframe_count)
					.css({"display": "inline-block",
						"overflow": "hidden",
						"background": figure_background});

				$(this).css({"opacity": 0,
					"display": "block"})
					.wrap(freezeframe_figure);

				freezeframe_figure = $(this).parent();

				// If an animation icon image is available, attach it
				if(_ff.animation_icon_image !== false) {
					var animation_icon = $("<div />")
						.attr("class", "freezeframe-animation-icon")
						.css({"display": "block",
							"position": "absolute",
							"background": "transparent " + "url('" + _ff.animation_icon_image + "') " +
								_ff.animation_icon_position + " no-repeat",
							"pointer-events": "none",
							"z-index": 100,
							"opacity": 0});

					$(freezeframe_figure).prepend(animation_icon);

					animation_icon = $(this).siblings($(".freezeframe-animation-icon"));
				}

				// Increment counter so each image gets a unique number
				_ff.freezeframe_count++;

				// Create a temporary refernce to the image as non-object to pass to .drawImage
				var _self = this;

				// Using imagesLoaded by Desandro because .load doesn't work on cached images
				$(this).imagesLoaded(function() {

					$(this).off("imagesLoaded");

					_ff.canvas[0].width = $(this)[0].clientWidth;
					_ff.canvas[0].height = $(this)[0].clientHeight;

					$(this).attr("animated", $(this).attr("src"))
						.attr("src", $(this).attr("src"));

					_ff.context.drawImage(_self, 0, 0, $(this)[0].clientWidth, $(this)[0].clientHeight);

					$(this).attr("src", _ff.canvas[0].toDataURL());

					// If an animation icon image is available, show it
					if(_ff.animation_icon_image !== false) {
						$(animation_icon).css({
							"width": parseInt($(this)[0].width),
							"height": parseInt($(this)[0].height)})
							.animate({"opacity": 1}, _ff.loading_fade_in_speed);
					}

					// When fade in sequence is complete, enable interaction
					$(this).animate({"opacity": 1}, _ff.loading_fade_in_speed, function() {

						$(freezeframe_figure).css("background", "url('" + $(this).attr("src") + "')");

						$(this).css("opacity", 0)
							.attr("src", $(this).attr("animated"));

						// Touch Device or Click Event
						if((_ff.support_touch_devices && _ff.is_touch_device) || _ff.trigger_event.toLowerCase() == "click") {

							var stop_animation;
							var animating = false;

							$(this).click(function() {

								// If not currently animating, start animating
								if(!animating) {
									$(this).attr("src", $(this).attr("src"))
										.css("opacity", 1);
									if(_ff.animation_icon_image !== false) {
										$(animation_icon).css("opacity", 0);
									}

									stop_animation = setInterval(function() {
										clearInterval(stop_animation);
										this.animate({"opacity": 0}, _ff.animation_fade_out_speed);
										if(_ff.animation_icon_image !== false) {
											$(animation_icon).animate({"opacity": 1}, _ff.animation_fade_out_speed);
										}
										animating = false;
									}, _ff.animation_play_duration);
									animating = true;

								// If currently animating, stop animating
								} else {
									clearInterval(stop_animation);
									$(this).animate({"opacity": 0}, _ff.animation_fade_out_speed);
									if(_ff.animation_icon_image !== false) {
										$(animation_icon).animate({"opacity": 1}, _ff.animation_fade_out_speed);
									}
									animating = false;
								}
							});

						} else {

							// Hover Event
							$(this).mouseenter(function() {
								$(this).attr("src", $(this).attr("src"))
									.css("opacity", 1);
								if(_ff.animation_icon_image !== false) {
									$(animation_icon).css("opacity", 0);
								}
							});

							$(this).mouseleave(function() {
								$(this).animate({"opacity": 0}, _ff.animation_fade_out_speed);
								if(_ff.animation_icon_image !== false) {
									$(animation_icon).animate({"opacity": 1}, _ff.animation_fade_out_speed);
								}
							});
						}
					});
				});
			}
		});
	};

	return FreezeFrame;
})(jQuery);

/** ----------------------------------------------------------------------------
 * Setup & Run Freezeframe
 * -------------------------------------------------------------------------- */

jQuery(document).ready(function($) {
	typeof(freezeframe_options) == 'undefined' ? freezeframe_options = {} : null;
	freezeframe = new FreezeFrame(freezeframe_options);
	freezeframe.setup();
	freezeframe.run();
});

/** ----------------------------------------------------------------------------
 * jQuery imagesLoaded plugin v2.1.1
 * http://github.com/desandro/imagesloaded
 *
 * MIT License. by Paul Irish et al.
 * -------------------------------------------------------------------------- */

;(function($, undefined) {
'use strict';
var BLANK = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==';
$.fn.imagesLoaded = function( callback ) {
	var $this = this,
		deferred = $.isFunction($.Deferred) ? $.Deferred() : 0,
		hasNotify = $.isFunction(deferred.notify),
		$images = $this.find('img').add( $this.filter('img') ),
		loaded = [],
		proper = [],
		broken = [];
	if ($.isPlainObject(callback)) {
		$.each(callback, function (key, value) {
			if (key === 'callback') {
				callback = value;
			} else if (deferred) {
				deferred[key](value);
			}
		});
	}
	function doneLoading() {
		var $proper = $(proper),
			$broken = $(broken);
		if ( deferred ) {
			if ( broken.length ) {
				deferred.reject( $images, $proper, $broken );
			} else {
				deferred.resolve( $images );
			}
		}
		if ( $.isFunction( callback ) ) {
			callback.call( $this, $images, $proper, $broken );
		}
	}
	function imgLoadedHandler( event ) {
		imgLoaded( event.target, event.type === 'error' );
	}
	function imgLoaded( img, isBroken ) {
		if ( img.src === BLANK || $.inArray( img, loaded ) !== -1 ) {
			return;
		}
		loaded.push( img );
		if ( isBroken ) {
			broken.push( img );
		} else {
			proper.push( img );
		}
		$.data( img, 'imagesLoaded', { isBroken: isBroken, src: img.src } );
		if ( hasNotify ) {
			deferred.notifyWith( $(img), [ isBroken, $images, $(proper), $(broken) ] );
		}
		if ( $images.length === loaded.length ) {
			setTimeout( doneLoading );
			$images.unbind( '.imagesLoaded', imgLoadedHandler );
		}
	}
	if ( !$images.length ) {
		doneLoading();
	} else {
		$images.bind( 'load.imagesLoaded error.imagesLoaded', imgLoadedHandler )
		.each( function( i, el ) {
			var src = el.src;
			var cached = $.data( el, 'imagesLoaded' );
			if ( cached && cached.src === src ) {
				imgLoaded( el, cached.isBroken );
				return;
			}
			if ( el.complete && el.naturalWidth !== undefined ) {
				imgLoaded( el, el.naturalWidth === 0 || el.naturalHeight === 0 );
				return;
			}
			if ( el.readyState || el.complete ) {
				el.src = BLANK;
				el.src = src;
			}
		});
	}
	return deferred ? deferred.promise( $this ) : $this;
};
})(jQuery);

/** ----------------------------------------------------------------------------
 * isMouseEventSupported by kangax
 * http://perfectionkills.com/detecting-event-support-without-browser-sniffing/
 * -------------------------------------------------------------------------- */

function isMouseEventSupported(eventName) {
  var el = document.createElement('div');
  eventName = 'on' + eventName;
  var isSupported = (eventName in el);
  if (!isSupported) {
    el.setAttribute(eventName, 'return;');
    isSupported = typeof el[eventName] == 'function';
  }
  el = null;
  return isSupported;
}
